reqlog/external_latency.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package reqlog
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// WithEnableExternalLatencyMeasurement returns a context that measures external latencies.
+func WithEnableExternalLatencyMeasurement(ctx context.Context) context.Context {
+	container := contextContainer{
+		latencies: make([]externalLatency, 0),
+	}
+	return context.WithValue(ctx, internalLatencyKey, &container)
+}
+
+// ExternalCallsMiddleware is a middleware that sets up the request context to measure external calls.
+// It has to be used before any other middleware that reads the final external latency.
+func ExternalCallsMiddleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	next(rw, r.WithContext(
+		WithEnableExternalLatencyMeasurement(r.Context()),
+	))
+}
+
+// MeasureExternalCall measures the duration of a function and records it as an external call.
+// The wrapped function's return value is returned.
+func MeasureExternalCall[T any](ctx context.Context, cause, detail string, f func() T) T {
+	defer StartMeasureExternalCall(ctx, cause, detail)()
+	return f()
+}
+
+// MeasureExternalCallErr measures the duration of a function and records it as an external call.
+// The wrapped function's return value and error is returned.
+func MeasureExternalCallErr[T any](ctx context.Context, cause, detail string, f func() (T, error)) (T, error) {
+	defer StartMeasureExternalCall(ctx, cause, detail)()
+	return f()
+}
+
+// StartMeasureExternalCall starts measuring the duration of an external call.
+// The returned function has to be called to record the duration.
+func StartMeasureExternalCall(ctx context.Context, cause, detail string) func() {
+	container, ok := ctx.Value(internalLatencyKey).(*contextContainer)
+	if !ok {
+		return func() {}
+	}
+
+	start := time.Now()
+	return func() {
+		container.Lock()
+		defer container.Unlock()
+		container.latencies = append(container.latencies, externalLatency{
+			Took:   time.Since(start),
+			Cause:  cause,
+			Detail: detail,
+		})
+	}
+}
```

Comment:
```suggestion
func MeasureExternalLatency(ctx context.Context, descriptor string, start time.Time) {
	container, ok := ctx.Value(internalLatencyKey).(*contextContainer)
	if !ok {
		return
	}

	container.Lock()
	defer container.Unlock()
	container.latencies = append(container.latencies, externalLatency{
		Took:   time.Since(start),
		Descriptor: descriptor,
	})
}
```